boards/support/tock_rubble/src/tock_rubble.rs

```diff
@@ -0,0 +1,449 @@
+where
+    A: time::Alarm<'a>,
+    R: hil::rubble::radio::RubbleData<'a>
+        + ble_advertising::BleAdvertisementDriver<'a>
+        + ble_advertising::BleConfig
+        + 'a,
+{
+    type Error = rubble::Error;
+
+    fn new(device_address: hil::rubble::types::DeviceAddress, alarm: &'a A) -> Self {
+        LinkLayerWrapper(rubble::link::LinkLayer::new(
+            rubble::link::DeviceAddress::new(
+                device_address.bytes,
+                match device_address.kind {
+                    hil::rubble::types::AddressKind::Public => rubble::link::AddressKind::Public,
+                    hil::rubble::types::AddressKind::Random => rubble::link::AddressKind::Random,
+                },
+            ),
+            timer_wrapper::TimerWrapper::new(alarm),
+        ))
+    }
+
+    fn start_advertise(
+        &mut self,
+        interval: hil::rubble::types::Duration,
+        data: &[u8],
+        transmitter: &mut BleRadioWrapper<'a, R>,
+        tx: refcell_packet_queue::RefCellConsumer<'static>,
+        rx: refcell_packet_queue::RefCellProducer<'static>,
+    ) -> Result<hil::rubble::types::NextUpdate, Self::Error> {
+        // This is inefficient, as we're parsing then directly turning back into
+        // bytes. However, this is a minimal-effort code to comply with Rubble's
+        // interface.
+
+        // A more efficient method might be to write a more minimal parser, and
+        // just directly create `AdStructure::Unknown` instances.
+        let mut ad_byte_reader = rubble::bytes::ByteReader::new(data);
+
+        // Note: advertising data is at most 31 octets, and each ad must be at
+        // least 2 octets, so we have at most 16 adstructures.
+        // TODO: allocating this on the stack might not be the most wise.
+        let mut ad_structures =
+            [rubble::link::ad_structure::AdStructure::Unknown { ty: 0, data: &[] }; 16];
+        let mut num_ads = 0;
+        while ad_byte_reader.bytes_left() > 0 {
+            let ad = rubble::link::ad_structure::AdStructure::from_bytes(&mut ad_byte_reader)?;
+            ad_structures[num_ads] = ad;
+            num_ads += 1;
+            if num_ads > 16 {
+                return Err(rubble::Error::InvalidLength);
+            }
+        }
+        let ad_structures = &ad_structures[..num_ads];
+
+        let res = self.0.start_advertise(
+            rubble::time::Duration::from_micros(interval.as_micros()),
+            ad_structures,
+            transmitter,
+            tx,
+            rx,
+        );
+        res.map(|next_update| match next_update {
+            rubble::link::NextUpdate::Keep => hil::rubble::types::NextUpdate::Keep,
+            rubble::link::NextUpdate::Disable => hil::rubble::types::NextUpdate::Disable,
+            rubble::link::NextUpdate::At(time) => hil::rubble::types::NextUpdate::At(
+                hil::rubble::types::Instant::from_raw_micros(time.raw_micros()),
+            ),
+        })
```

Comment:
I don't understand why the tock-rubble layer pushes the timer loop logic back down to the caller (in this case the Driver capsule). It seems like this interface layer should have to handle managing the timers for the Rubble library, and not push them though making every kernel use of Rubble have to deal with it.

Reply:
I'm not sure what you mean here, can you point to the code?